### ISSUE TITLE
Inconsistent results for epsilon * inf on CPU

### ISSUE THREAD

**Issue by mathisgerdes** [google/jax#4042]
To be consistent with numpy, infinity multiplied by small positive values should return infinity.
For small enough values, jax returns `nan` instead.

```python
>>> import jax.numpy as jnp
>>> import numpy as onp

>>> onp.nextafter(0, 1) * onp.inf
inf

>>> jnp.finfo(jnp.float_).eps * jnp.inf  # correct for some small values
inf

>>> jnp.nextafter(0, 1)  # smaller than eps above, but != 0
DeviceArray(1.e-45, dtype=float32)

>>> jnp.nextafter(0, 1) * jnp.inf  # should also return inf
DeviceArray(nan, dtype=float32)
```

I've tried this on my local machine and on Colab running on the CPU with 32 and 64 bit versions, jax version `0.1.75`. Running on GPU it correctly returns `inf`.

Curiously, on the CPU in 32-bit mode (but not on the GPU on Colab) the following returns true
```python
jnp.nextafter(0, 1) == onp.nextafter(0, 1)
```
despite ostensibly being `1e-45` and `5e-324`, respectively.

**Comment by hawkinsp** [comment]
I think the issue is that `jnp.nextafter` is returning a denormal number, but in general JAX has flush-denormal-to-zero semantics on CPU. Denormals significantly slow floating-point code on CPU and we do not intend to allow them. Perhaps we should change `jnp.nextafter` to return the smallest normalized number in this case.

I'm guessing that GPU does not have flush-denormal-to-zero semantics enabled. We could perhaps enable them there also for consistency, but I am not sure there is any performance benefit to doing so.

The difference between `onp.nextafter` and `jnp.nextafter` is probably because `onp.nextafter` is returning a float64 not a float32. `5e-324` is not a representable `float32`.

What do you think?

**Comment by mathisgerdes** [comment]
I don't think `jnp.nextafter` should return different values depending on whether CPU or GPU is used. Even though the value is effectively useless on the CPU (since flushed to zero in any operation) it is evidently representable, so I think the current behavior is the correct one. Maybe this is a subtlety that should be mentioned in the docstring of `jnp.nextafter` (i.e. the fact that potentially non-usable denormal numbers are returned).

The consideration might be different if denormal numbers were in fact not representable when using flush-dernormal-to-zero semantics, since then `np.nextafter(0, 1)` shouldn't return `0` on the CPU.

In my case the issue was that I wasn't aware of the flush-denormal-to-zero semantics in JAX. Once the denormal numbers are identified with zero the behavior makes sense, and instead of manually constructing the epsilon around `0` one should use `jnp.finfo(jnp.float_).tiny` (the smallest normal float; although it is a numpy object at first on which operations don't flush denormal values to `0`).

> The difference between onp.nextafter and jnp.nextafter is probably because onp.nextafter is returning a float64 not a float32. 5e-324 is not a representable float32.

I suspect that in the comparison also flush-dernormal-to-zero occurs, thus leading to `0 == 0`. On the GPU it doesn't, giving the (more expected) result of `1e-45 == 0 (= float32(5e-324))`.
